osm2lanes/src/transform/tags_to_lanes/modes/foot_shoulder.rs

```diff
@@ -37,107 +37,126 @@ impl LaneBuilder {
 
-            (Some("separate"), None) => (Sidewalk::Separate, Sidewalk::No),
-            (None, Some("separate")) => (Sidewalk::No, Sidewalk::Separate),
-            (Some(_), None) | (None, Some(_)) | (Some(_), Some(_)) => {
+                (Some("separate"), None) => (Sidewalk::Separate, Sidewalk::No),
+                (None, Some("separate")) => (Sidewalk::No, Sidewalk::Separate),
+                (Some(_), None) | (None, Some(_)) | (Some(_), Some(_)) => {
+                    return err;
+                },
+            },
+            (Some(_), Some(_), (_, _))
+            | (Some(_), _, (_, Some(_)) | (Some(_), _))
+            | (_, Some(_), (_, Some(_)) | (Some(_), _)) => {
                 return err;
             },
-        },
-        (Some(_), Some(_), (_, _))
-        | (Some(_), _, (_, Some(_)) | (Some(_), _))
-        | (_, Some(_), (_, Some(_)) | (Some(_), _)) => {
-            return err;
-        },
-    };
+        };
+        Ok(sidewalks)
+    }
+}
+
+enum Shoulder {
+    None,
+    Yes,
+    No,
+}
+
+impl Shoulder {
+    fn from_tags(
+        tags: &Tags,
+        locale: &Locale,
+        _warnings: &mut RoadWarnings,
+    ) -> Result<(Self, Self), TagsToLanesMsg> {
+        Ok(match tags.get(SHOULDER) {
+            None => (Shoulder::None, Shoulder::None),
+            Some("no") => (Shoulder::No, Shoulder::No),
+            Some("yes" | "both") => (Shoulder::Yes, Shoulder::Yes),
+            Some(s) if s == locale.driving_side.tag().as_str() => (Shoulder::Yes, Shoulder::No),
+            Some(s) if s == locale.driving_side.opposite().tag().as_str() => {
+                (Shoulder::No, Shoulder::Yes)
+            },
+            Some(s) => return Err(TagsToLanesMsg::unsupported_tag(SHOULDER, s)),
+        })
+    }
+}
+
+#[allow(clippy::items_after_statements, clippy::unnested_or_patterns)]
+pub(in crate::transform::tags_to_lanes) fn foot_and_shoulder(
+    tags: &Tags,
+    locale: &Locale,
+    road: &mut RoadBuilder,
+    warnings: &mut RoadWarnings,
+) -> Result<(), RoadError> {
+    // https://wiki.openstreetmap.org/wiki/Key:sidewalk
+    let sidewalk: (Sidewalk, Sidewalk) = Sidewalk::from_tags(tags, locale, warnings)?;
 
     // https://wiki.openstreetmap.org/wiki/Key:shoulder
-    enum Shoulder {
-        None,
-        Yes,
-        No,
-    }
-    let shoulder: (Shoulder, Shoulder) = match tags.get(SHOULDER) {
-        None => (Shoulder::None, Shoulder::None),
-        Some("no") => (Shoulder::No, Shoulder::No),
-        Some("yes" | "both") => (Shoulder::Yes, Shoulder::Yes),
-        Some(s) if s == locale.driving_side.tag().as_str() => (Shoulder::Yes, Shoulder::No),
-        Some(s) if s == locale.driving_side.opposite().tag().as_str() => {
-            (Shoulder::No, Shoulder::Yes)
-        },
-        Some(s) => return Err(TagsToLanesMsg::unsupported_tag(SHOULDER, s).into()),
-    };
+    let shoulder: (Shoulder, Shoulder) = Shoulder::from_tags(tags, locale, warnings)?;
```

Comment:
this is a mostly a refactor

I only changed the return  Err from `RoadError` to `TagsToLanesMsg`